### ISSUE TITLE
Flaky test: `Application_VisualStyleState_Set_ReturnsExpected` deadlock

### ISSUE THREAD

**Issue by RussKie** [dotnet/winforms#2192]
**Problem description:**

As part of tests deadlock investigation, the major deadlock causing issue was eliminated in #2191, however that uncovered a secondary issue that `MonthCalendar_CalendarDimensions_SetAreaOfTwelve_GetReturnsExpected` could deadlock as well.

**Actual behavior:** 

![image](https://user-images.githubusercontent.com/4403806/67532139-e3821000-f710-11e9-83c8-d0352cc71b4e.png)

All but two threads appear to be waiting.

This thread is executing out code appears to be stuck sending "theme changed" messages
```
Flagged	>	71456	11	Worker Thread	<No Name>	System.Windows.Forms.dll!System.Windows.Forms.Application.SendThemeChangedRecursive

 	[Managed to Native Transition]		Annotated Frame
>	System.Windows.Forms.dll!System.Windows.Forms.Application.SendThemeChangedRecursive(System.IntPtr handle = 0x00000000000727ca, System.IntPtr lparam = 0x0000000000000000) Line 648	C#	Symbols loaded.
 	[Native to Managed Transition]		Annotated Frame
 	[Managed to Native Transition]		Annotated Frame
 	System.Windows.Forms.dll!System.Windows.Forms.Application.SendThemeChangedRecursive(System.IntPtr handle = 0x0000000000971f24, System.IntPtr lparam = 0x0000000000000000) Line 643	C#	Symbols loaded.
 	[Native to Managed Transition]		Annotated Frame
 	[Managed to Native Transition]		Annotated Frame
 	System.Windows.Forms.dll!System.Windows.Forms.Application.SendThemeChangedRecursive(System.IntPtr handle = 0x00000000002320fe, System.IntPtr lparam = 0x0000000000000000) Line 643	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Application.SendThemeChanged(System.IntPtr handle = 0x00000000002320fe, System.IntPtr extraParameter = 0x0000000000000000) Line 625	C#	Symbols loaded.
 	[Native to Managed Transition]		Annotated Frame
 	[Managed to Native Transition]		Annotated Frame
 	System.Windows.Forms.dll!System.Windows.Forms.Application.VisualStyleState.set(System.Windows.Forms.VisualStyles.VisualStyleState value = NoneEnabled) Line 609	C#	Symbols loaded.
 	System.Windows.Forms.Tests.dll!System.Windows.Forms.Tests.ApplicationTests.Application_VisualStyleState_Set_ReturnsExpected(System.Windows.Forms.VisualStyles.VisualStyleState value = NoneEnabled) Line 43	C#	Symbols loaded.
 	[Native to Managed Transition]		Annotated Frame
 	[Managed to Native Transition]		Annotated Frame
 	xunit.execution.dotnet.dll!Xunit.Sdk.TestInvoker<System.__Canon>.CallTestMethod(object testClassInstance = {System.Windows.Forms.Tests.ApplicationTests}) Line 150	C#	Non-user code. Symbols loaded.

```
And this thread I presume responsible for handling events?
```
Flagged		87564	22	Worker Thread	.NET SystemEvents	Microsoft.Win32.SystemEvents.dll!Microsoft.Win32.SystemEvents.WindowThreadProc

 	[Managed to Native Transition]		Annotated Frame
>	Microsoft.Win32.SystemEvents.dll!Microsoft.Win32.SystemEvents.WindowThreadProc()	Unknown	Non-user code. Skipped loading symbols.
 	System.Private.CoreLib.dll!System.Threading.ThreadHelper.ThreadStart_Context(object state)	Unknown	Non-user code. Skipped loading symbols.
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state)	Unknown	Non-user code. Skipped loading symbols.
 	System.Private.CoreLib.dll!System.Threading.ThreadHelper.ThreadStart()	Unknown	Non-user code. Skipped loading symbols.

```

**Expected behavior:**

Tests don't deadlock

**Minimal repro:**

Uncomment tests and run `build -test`

**Comment by weltkante** [comment]
What is happening here is that this test is testing something that needs every single control and window handle to be operational, including those from other tests. Changing the theme state enumerates all windows and sends them a WM_THEMECHANGED (that part is in the screenshot). If there is a single window/control which isn't responding then this test will hang.

A control will not respond if its owning thread isn't running a message loop currently. Normally this happens when the UI thread is busy doing some non-UI task but that shouldn't be the problem here. I think its more likely that some test created a control handle on a thread which isn't supposed to have controls. Once the test completed nobody cleaned up the handle and its a leftover. Since the test thread isn't running a message loop the control will never respond nor be cleaned up. (GC just marks the control for cleanup but without a message loop it will never happen.)

Basically what this test failure is saying is: _some other test messed up_.

I think its acceptable to leave this test off for now, identifying the offending test(s) might be hard. If you want to approach this structurally I'd suggest checking after every non-UI test that there were no handles accidentally created ([EnumThreadWindows](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-enumthreadwindows) on the current thread). This should probably be a diagnostic mode and not run on normal testing.

**Comment by RussKie** [comment]
Thank you for the analysis. As usual, very insightful and helpful 💯

**Comment by weltkante** [comment]
> Furthermore, at the point of the deadlock there are a whole bunch of forms open even though no other test is running any more

When executing all tests from command line this is no longer happening, forms are now closing as expected. Might be possible to reenable this test, if it still hangs its probably a different reason now.

**Comment by hughbe** [comment]
I would suggest putting this in a `RemoteExecutor.Invoke` call as well, just to be safe

**Comment by RussKie** [comment]
> Might be possible to reenable this test, if it still hangs its probably a different reason now.

I think we have a problem with the RE now - https://github.com/dotnet/arcade/issues/5325
But, yes, I think we may be in a position to start re-enabling tests and see what gives.
